### ISSUE TITLE
RML Function Cannot Be Found

### ISSUE THREAD

**Issue by tobiasschweizer** [carml/carml#270]
Hi there :-)

I am trying to apply the sum function:

```ttl

PREFIX rr: <http://www.w3.org/ns/r2rml#>
PREFIX rml: <http://semweb.mmlab.be/ns/rml#>
PREFIX ql: <http://semweb.mmlab.be/ns/ql#>
PREFIX carml: <http://carml.taxonic.com/carml/>
PREFIX dct: <http://purl.org/dc/terms/>
PREFIX premis: <http://id.loc.gov/vocabulary/preservation/>
PREFIX schema: <http://schema.org/>
PREFIX fno: <https://w3id.org/function/ontology#>
PREFIX fnml: <http://semweb.mmlab.be/ns/fnml#>
PREFIX grel: <http://users.ugent.be/~bjdmeest/function/grel.ttl#>
PREFIX ex: <http://example.org/>
@base <http://example.com/ns#>.

...


  rr:predicateObjectMap [
    rr:predicate schema:value    ;
    rr:objectMap [
      a fnml:FunctionMap ;
      fnml:functionValue [
        rml:logicalSource <#LogicalSourceGrant> ;
        rr:subjectMap [
          rr:template "functionExec";
          rr:termType rr:BlankNode ;
          rr:class fno:Execution
        ] ;
        rr:predicateObjectMap
          [
            rr:predicate fno:executes ;
            rr:objectMap [
              rr:constant ex:sumFunction ;
            ]
          ] ,
          [
            rr:predicate ex:intParameterA ;
            rr:objectMap [ rml:reference "AmountGrantedAllSets" ]
          ] ,
          [
            rr:predicate ex:intParameterB  ;
            rr:objectMap [ rml:reference "AmountGrantedAllSets" ]
          ]
      ] ;
      rr:datatype xsd:integer ;
    ]
  ] .
```

I get 

> io.carml.engine.TermGeneratorFactoryException: no function registered for function IRI [http://example.org/sumFunction]

Could you tell me what I am doing wrong? 

As far as I can see, the function is defined and implemented here: https://github.com/carml/carml/blob/d19e44bf60b51665c343bd823829bd1fd708be2c/carml-engine/src/test/java/io/carml/engine/iotests/RmlFunctions.java#L67

Actually, what I need are two functions to:
- split a string by semicolon (to split keywords)
- strip day time from a date (to get day precision)

I am not sure if they are provided by FnO. 

Thanks and kind regards,

Tobias

**Comment by pmaria** [comment]
Hi @tobiasschweizer,

I'm afraid CARML doesn't ship with any registered functions by default, but expects them to be registered via the builder `mapperBuilder.addFunctions(..)`. 


> As far as I can see, the function is defined and implemented here:
> 
> https://github.com/carml/carml/blob/d19e44bf60b51665c343bd823829bd1fd708be2c/carml-engine/src/test/java/io/carml/engine/iotests/RmlFunctions.java#L67

That is just a test, and not part of the main artifact.


> Actually, what I need are two functions to:
> 
> * split a string by semicolon (to split keywords)
> * strip day time from a date (to get day precision)
> 
> I am not sure if they are provided by FnO.

So currently, to use functions, you will have to define them manually as described in https://github.com/carml/carml#function-extension.

Hope this helps.

**Comment by tobiasschweizer** [comment]
Ok, that is pretty neat. I have just tested the implementation of a split function ...
Handling that outside of RML is really cumbersome ...

```java
@FnoFunction("http://example.org/splitFunction")
    public List<String> splitFunction(
            @FnoParam("http://example.org/strParameterA") String strA
    ) {
        System.out.println(strA);
        return Arrays.asList(strA.split(";"));
    }
```

**Comment by pmaria** [comment]
> I understand that FnO specifies how to declare an RML function and formally describe its parameters and result. However, I also understood that FnO aims at achieving [implementation-independence](https://imec-publications.be/handle/20.500.12860/39889), so I wonder how the functions could be implemented in a way so that they could be used with different RML engines.

Agreed the implementation independence needs work. Since there is a new emerging specification for RML including functions, it makes sense to look into how we can get there.

**Comment by tobiasschweizer** [comment]
Great. Maybe that would be a thing I could help with since it seems quite modular. Is this the latest spec https://rml.io/specs/rml/?

In general, I would like to adopt RML functions step by step to reduce pre-processing.

**Comment by tobiasschweizer** [comment]
Ok, I posted my question here: https://github.com/kg-construct/rml-fnml/issues/28. This can be closed then.
